### PR TITLE
Add feature to configure DI via the config

### PR DESCRIPTION
**Pull request by tomphp**

### CODE REVIEW

File: src/DIConfigServiceProvider.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    private function registerService($name, array $config)
+    {
+        $singleton = array_key_exists('singleton', $config) && $config['singleton'];
+
+        $service = $this->getContainer()->add($name, $config['class'], $singleton);
+
+        if (!$service) {
+            return;
+        }
```
Comment (robteifi) [inline src/DIConfigServiceProvider.php]: Possibly throw?

File: src/JSONFileReader.php [hunk 2]
```diff
@@ -29,7 +29,7 @@ public function read($filename)
     private function assertFileExists()
     {
         if (!file_exists($this->filename)) {
-            throw new FileNotFoundException("{$this->filename} does not exist");
+            throw new FileNotFoundException($this->filename . ' does not exist');
```
Comment (localheinz) [inline src/JSONFileReader.php]: @tomphp 

Maybe consistently use `sprintf()`?

File: src/ReaderFactory.php [hunk 3]
```diff
@@ -57,10 +57,11 @@ private function getReaderClass($filename)
         }
 
         if (!$found) {
-            throw new UnknownFileTypeException(
-                "Not reader class found for $filename; configured exceptions are "
-                . implode(', ', array_keys($this->config))
-            );
+            throw new UnknownFileTypeException(sprintf(
+                'Not reader class found for %s; configured exceptions are %s',
```
Comment (localheinz) [inline src/ReaderFactory.php]: @tomphp 

Should probably be `No reader class`, shouldn't it?

File: tests/DIConfigServiceProviderTest.php [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+        $this->container->addServiceProvider(new DIConfigServiceProvider($config));
+
+        $instance1 = $this->container->get('example_class');
+        $instance2 = $this->container->get('example_class');
+
+        $this->assertNotSame($instance1, $instance2);
+    }
+
+    public function testItDefaultsToCreatingUniqueInstances()
```
Comment (localheinz) [inline tests/DIConfigServiceProviderTest.php]: @tomphp 

Maybe a bit ambiguous, as in fact, it does not provide the same instance, but a new one every time?

**Comment by localheinz** [comment]
@tomphp 

Looks good, and restores the functionality from `1.0`, so :+1:!

**Pull request merged**